### PR TITLE
LF: Use template parameter (`this`) in method bodies.

### LINKED ISSUE #13123: Reuse bound `this` in template implements method expression.
Currently in LF, the "template implements method" value field is a function `templateType -> methodType`. But templates have an explicit (this: templateType) binder, which we should reuse. So the idea is to change the field to be an expression of type `methodType` and reuse `(this: templateType)` in the context. This would save a bit of space and make the AST a bit more internally consistent.

Making this change involves updating both type checkers, LFConversion, Speedy, and the relevant parts of the LF spec (the typing rules and operational semantics).

### PR DESCRIPTION
**Pull request by sofiafaro-da**
Instead of having method bodies be functions from the template type to the method type, have them just be expressions of the method type that reuse the existing template parameter (`this`).

Fixes #13123

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs [hunk 1]
```diff
@@ -962,8 +961,8 @@ checkFeature feature = do
 
 checkTemplateKey :: MonadGamma m => ExprVarName -> Qualified TypeConName -> TemplateKey -> m ()
 checkTemplateKey param tcon TemplateKey{..} = do
+    checkType tplKeyType KStar
```
Comment (sofiafaro-da) [inline compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/Check.hs:964]: Unrelated change: the key type should be typechecked in the empty context. This matches the behavior of the scala typechecker and the spec.

File: daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala [hunk 2]
```diff
@@ -443,13 +443,13 @@ private[validation] object Typing {
...
       env.checkExpr(agreementText, TText)
       choices.values.foreach(env.checkChoice(tplName, _))
+      env.checkIfaceImplementations(tplName, implementations)
       mbKey.foreach { key =>
         checkType(key.typ, KStar)
         env.checkExpr(key.body, key.typ)
         checkExpr(key.maintainers, TFun(key.typ, TParties))
         ()
       }
-      env.checkIfaceImplementations(tplName, implementations)
```
Comment (sofiafaro-da) [inline daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala:452]: This was buggy -- the implementation had access to a `this` param when it shouldn't have.
Now it should have access to it. 

(I changed the order here so that it matches the Haskell side.)

**Review by akrmn — APPROVED** [review]
Looks good, thank you!

**Review by cocreature — APPROVED** [review]
Nice, thanks!

**Review by robin-da — APPROVED** [review]
Looks good, thanks!
